File: backend/app/utils/credential_encryption.py
"""AES-256-GCM encryption helpers for sensitive credentials stored at rest.

Format: ``v1:<base64(12-byte-nonce || ciphertext || 16-byte-tag)>``

Usage::

    from app.utils.credential_encryption import encrypt, decrypt, is_encrypted

    stored = encrypt("my-password", secret_key="deployment-secret")
    plaintext = decrypt(stored, secret_key="deployment-secret")

When ``secret_key`` is empty the helpers fall back to plaintext with a
warning, so the app works out-of-the-box without requiring extra config.
"""

from __future__ import annotations

import base64
import hashlib
import logging
import os

logger = logging.getLogger(__name__)

_PREFIX = "v1:"
_NONCE_LEN = 12
_TAG_LEN = 16
_AAD = b"smtp"  # additional authenticated data — binds ciphertext to context


def _derive_key(secret: str) -> bytes:
    """Derive a 32-byte AES key from the deployment secret via SHA-256."""
    return hashlib.sha256(secret.encode()).digest()


def encrypt(plaintext: str, secret_key: str) -> str:
    """Encrypt *plaintext* and return a ``v1:…`` ciphertext string.

    Falls back to returning *plaintext* unchanged (with a warning) when
    *secret_key* is empty so the application works without extra config.
    """
    if not secret_key:
        logger.warning(
            "DRIVECHILL_SECRET_KEY is not set; SMTP password stored in plaintext. "
            "Set the env var to enable at-rest encryption."
        )
        return plaintext

    from cryptography.hazmat.primitives.ciphers.aead import AESGCM

    nonce = os.urandom(_NONCE_LEN)
    aesgcm = AESGCM(_derive_key(secret_key))
    # AESGCM.encrypt returns ciphertext + tag concatenated (tag appended at end)
    ct_and_tag = aesgcm.encrypt(nonce, plaintext.encode(), _AAD)
    payload = base64.b64encode(nonce + ct_and_tag).decode()
    return _PREFIX + payload


def decrypt(stored: str, secret_key: str) -> str:
    """Decrypt a ``v1:…`` ciphertext string and return the plaintext.

    Returns *stored* unchanged if it is not encrypted (plaintext fallback).
    Returns an empty string when the key is missing but the value is encrypted.
    """
    if not stored.startswith(_PREFIX):
        return stored  # plaintext — no migration needed yet

    if not secret_key:
        logger.warning(
            "DRIVECHILL_SECRET_KEY is not set but the stored SMTP password is "
            "encrypted. Returning empty password; configure the secret key to "
            "decrypt credentials."
        )
        return ""

    from cryptography.hazmat.primitives.ciphers.aead import AESGCM

    try:
        data = base64.b64decode(stored[len(_PREFIX):])
        nonce = data[:_NONCE_LEN]
        ct_and_tag = data[_NONCE_LEN:]  # cryptography lib separates tag internally
        aesgcm = AESGCM(_derive_key(secret_key))
        return aesgcm.decrypt(nonce, ct_and_tag, _AAD).decode()
    except Exception:
        logger.exception("Failed to decrypt stored credential — returning empty string")
        return ""


def is_encrypted(value: str) -> bool:
    """Return True if *value* looks like an encrypted credential."""
    return value.startswith(_PREFIX)

File: backend-cs/Utils/CredentialEncryption.cs
using System.Security.Cryptography;
using System.Text;

namespace DriveChill.Utils;

/// <summary>
/// AES-256-GCM encryption helpers for sensitive credentials stored at rest.
///
/// Format: <c>v1:&lt;base64(12-byte-nonce || ciphertext || 16-byte-tag)&gt;</c>
///
/// When <c>secretKey</c> is null or empty the helpers fall back to plaintext
/// with a log warning so the application works without extra configuration.
/// </summary>
public static class CredentialEncryption
{
    private const string Prefix  = "v1:";
    private const int    NonceLen = 12;
    private const int    TagLen   = 16;
    private static readonly byte[] Aad = "smtp"u8.ToArray();

    /// <summary>
    /// Derive a 256-bit key from the deployment secret.
    /// IMPORTANT: SecretKey MUST be high-entropy (e.g. <c>secrets.token_hex(32)</c>).
    /// A single SHA-256 pass is acceptable for a 256-bit random input; a low-entropy
    /// passphrase would be brute-forceable. See AppSettings.SecretKey docs.
    /// </summary>
    private static byte[] DeriveKey(string secretKey)
        => SHA256.HashData(Encoding.UTF8.GetBytes(secretKey));

    /// <summary>
    /// Encrypt <paramref name="plaintext"/> and return a <c>v1:…</c> ciphertext string.
    /// Returns <paramref name="plaintext"/> unchanged (with a warning) when
    /// <paramref name="secretKey"/> is null or empty.
    /// </summary>
    public static string Encrypt(string plaintext, string? secretKey)
    {
        if (string.IsNullOrEmpty(secretKey))
            return plaintext;   // plaintext fallback — caller should log a warning

        var key   = DeriveKey(secretKey);
        var nonce = new byte[NonceLen];
        RandomNumberGenerator.Fill(nonce);

        var pt  = Encoding.UTF8.GetBytes(plaintext);
        var ct  = new byte[pt.Length];
        var tag = new byte[TagLen];

        using var aes = new AesGcm(key, TagLen);
        aes.Encrypt(nonce, pt, ct, tag, Aad);

        var payload = new byte[NonceLen + ct.Length + TagLen];
        nonce.CopyTo(payload, 0);
        ct.CopyTo(payload, NonceLen);
        tag.CopyTo(payload, NonceLen + ct.Length);

        return Prefix + Convert.ToBase64String(payload);
    }

    /// <summary>
    /// Decrypt a <c>v1:…</c> ciphertext string and return the plaintext.
    /// Returns <paramref name="stored"/> unchanged if it is not encrypted.
    /// Returns an empty string when the key is missing but the value is encrypted.
    /// </summary>
    public static string Decrypt(string stored, string? secretKey)
    {
        if (!stored.StartsWith(Prefix, StringComparison.Ordinal))
            return stored;   // plaintext fallback

        if (string.IsNullOrEmpty(secretKey))
            return string.Empty;   // can't decrypt without key

        try
        {
            var data  = Convert.FromBase64String(stored[Prefix.Length..]);
            var nonce = data[..NonceLen];
            var tag   = data[^TagLen..];
            var ct    = data[NonceLen..^TagLen];
            var pt    = new byte[ct.Length];

            var key = DeriveKey(secretKey);
            using var aes = new AesGcm(key, TagLen);
            aes.Decrypt(nonce, ct, tag, pt, Aad);
            return Encoding.UTF8.GetString(pt);
        }
        catch (CryptographicException)
        {
            // Wrong key or tampered ciphertext — caller gets empty string.
            return string.Empty;
        }
    }

    /// <summary>Returns true if <paramref name="value"/> is an encrypted credential.</summary>
    public static bool IsEncrypted(string value)
        => value.StartsWith(Prefix, StringComparison.Ordinal);
}
